manager/eris-mint/evm/snative.go

```diff
@@ -1,104 +1,351 @@
 	F        NativeContract
 }
 
-/* The solidity interface used to generate the abi function ids below
-contract Permissions {
-	function has_base(address addr, uint64 permFlag) constant returns (bool value) {}
-	function set_base(address addr, uint64 permFlag, bool value) constant returns (bool val) {}
-	function unset_base(address addr, uint64 permFlag) constant returns (uint64 pf) {}
-	function set_global(uint64 permFlag, bool value) constant returns (uint64 pf) {}
-	function has_role(address addr, string role) constant returns (bool val) {}
-	function add_role(address addr, string role) constant returns (bool added) {}
-	function rm_role(address addr, string role) constant returns (bool removed) {}
+type SolidityType string
+
+type SolidityArg struct {
+	Name string
+	Type SolidityType
 }
-*/
 
-// function identifiers from the solidity abi
-var PermsMap = map[string]SNativeFuncDescription{
-	getFuncIdentifiersFromSignature("has_role(address,bytes32)"):    SNativeFuncDescription{"has_role", 2, ptypes.HasRole, has_role},
-	getFuncIdentifiersFromSignature("unset_base(address,uint64)"):   SNativeFuncDescription{"unset_base", 2, ptypes.UnsetBase, unset_base},
-	getFuncIdentifiersFromSignature("set_global(uint64,bool)"):      SNativeFuncDescription{"set_global", 2, ptypes.SetGlobal, set_global},
-	getFuncIdentifiersFromSignature("add_role(address,bytes32)"):    SNativeFuncDescription{"add_role", 2, ptypes.AddRole, add_role},
-	getFuncIdentifiersFromSignature("set_base(address,uint64,bool"): SNativeFuncDescription{"set_base", 3, ptypes.SetBase, set_base},
-	getFuncIdentifiersFromSignature("has_base(address,uint64)"):     SNativeFuncDescription{"has_base", 2, ptypes.HasBase, has_base},
-	getFuncIdentifiersFromSignature("rm_role(address,bytes32)"):     SNativeFuncDescription{"rm_role", 2, ptypes.RmRole, rm_role},
+type SolidityReturn struct {
+	Name string
+	Type SolidityType
 }
 
-func getFuncIdentifiersFromSignature(signature string) string {
-	identifier := sha3.Sha3([]byte(signature))
-	return hex.EncodeToString(identifier[:4])
+const (
+	// We don't need to be exhaustive here, just make what we used strongly typed
+	SolidityAddress SolidityType = "address"
+	SolidityInt     SolidityType = "int"
+	SolidityUint64  SolidityType = "uint64"
+	SolidityBytes32 SolidityType = "bytes32"
+	SolidityString  SolidityType = "string"
+	SolidityBool    SolidityType = "bool"
+)
+
+func registerSNativeContracts() {
+	for _, contract := range SNativeContracts() {
+		registeredNativeContracts[contract.Address()] = contract.Dispatch
+	}
 }
 
-func permissionsContract(appState AppState, caller *Account, args []byte, gas *int64) (output []byte, err error) {
-	if len(args) < 4 {
-		return nil, fmt.Errorf("permissionsContract expects at least a 4-byte function identifier")
+// Returns a map of all SNative contracts defined indexed by name
+func SNativeContracts() map[string]SNativeContractDescription {
+	permFlagType := SolidityUint64
+	roleType := SolidityBytes32
```

Comment:
Internally we treat roles as strings, but I agree that it is better and sufficient to type it strictly as `bytes32`; alternatively we should change our impl to treat it as `Word256`